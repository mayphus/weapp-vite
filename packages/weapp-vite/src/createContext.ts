import type { LoadConfigOptions } from './context'
import { getCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const ctx = getCompilerContext(options?.key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
