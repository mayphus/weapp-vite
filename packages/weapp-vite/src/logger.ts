import logger from '@weapp-core/logger'

export default logger

export type LogType = 'error' | 'warn' | 'info'
export type LogLevel = LogType | 'silent'
