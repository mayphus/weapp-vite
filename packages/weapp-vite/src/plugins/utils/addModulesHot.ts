import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
