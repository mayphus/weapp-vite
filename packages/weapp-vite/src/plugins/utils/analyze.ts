import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
