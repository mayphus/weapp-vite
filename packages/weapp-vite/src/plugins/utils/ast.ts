import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
