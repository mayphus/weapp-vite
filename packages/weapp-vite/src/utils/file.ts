import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
