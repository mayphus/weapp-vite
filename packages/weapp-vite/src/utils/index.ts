export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
