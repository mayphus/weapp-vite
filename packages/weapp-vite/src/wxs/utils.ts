import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
