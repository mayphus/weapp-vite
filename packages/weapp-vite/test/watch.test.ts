import { createCompilerContext } from '@/createContext'
import logger from '@/logger'
import { sort } from 'fast-sort'
import fs from 'fs-extra'
import path from 'pathe'
// import whyIsNodeRunning from 'why-is-node-running'
import { getFixture, scanFiles } from './utils'

vi.mock('@/logger', async (importOriginal) => {
  const mod = await importOriginal()
  // @ts-ignore
  mod.default.success = vi.fn()
  return {
    // @ts-ignore
    ...mod,
  }
})

describe('watch', () => {
  it('watch', async () => {
    const cwd = getFixture('watch')
    const distDir = path.resolve(cwd, 'dist')
    await fs.remove(distDir)
    let subWatchFiles: string[] = []
    let rootWatchFiles: string[] = []

    const resolveIdFiles: string[] = []
    const loadFiles: string[] = []
    const subResolveIdFiles: string[] = []
    const subLoadFiles: string[] = []

    function resolveAbsPath(files: string[], sorted: boolean = true) {
      const x = files
        .filter(x => !x.includes('node_modules'))
        .map((x) => {
          return ctx.configService.relativeAbsoluteSrcRoot(x)
        })

      return sorted ? x.sort() : x
    }

    const ctx = await createCompilerContext({
      cwd,
      isDev: true,
      inlineConfig: {
        weapp: {
          debug: {
            watchFiles(watchFiles, meta) {
              if (meta) {
                subWatchFiles = resolveAbsPath(watchFiles)

                expect(
                  subWatchFiles,
                ).toMatchSnapshot(`watchFiles-${meta.subPackage.root}`)
              }
              else {
                rootWatchFiles = resolveAbsPath(watchFiles)
                expect(
                  rootWatchFiles,
                ).toMatchSnapshot('watchFiles')
              }
            },
            resolveId(id, meta) {
              if (meta) {
                subResolveIdFiles.push(id)
              }
              else {
                resolveIdFiles.push(id)
              }
            },
            load(id, meta) {
              if (meta) {
                subLoadFiles.push(id)
              }
              else {
                loadFiles.push(id)
              }
            },
          },
        },
      },
    })
    await ctx.buildService.build()
    expect(await fs.exists(distDir)).toBe(true)
    const files = await scanFiles(distDir)
    expect(files).toMatchSnapshot()

    expect(ctx.scanService.subPackageMap).toMatchSnapshot()
    expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm'))).toBe(true)
    expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/buffer'))).toBe(true)
    expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/gm-crypto'))).toBe(true)

    expect(logger.success).toHaveBeenNthCalledWith(
      1,
      '已清空 dist/ 目录',
    )
    // expect(logger.success).toHaveBeenNthCalledWith(
    //   2,
    //   '[npm] `buffer/` 依赖处理完成!',
    // )
    // expect(logger.success).toHaveBeenNthCalledWith(
    //   3,
    //   '[npm] 分包[packageB] `buffer/` 依赖处理完成!',
    // )

    expect(sort(
      await Promise.all(files.filter(x => x.endsWith('.wxs')).map(async (x) => {
        return {
          file: x,
          code: await fs.readFile(path.resolve(ctx.configService.outDir, x), 'utf8'),
        }
      })),
    ).by([
      {
        asc: 'file',
      },
    ])).toMatchSnapshot('wxs')

    // expect(resolveAbsPath(resolveIdFiles)).toMatchSnapshot('resolveIdFiles')
    // expect(resolveAbsPath(loadFiles, false)).toMatchSnapshot('loadFiles')

    // expect(resolveAbsPath(subResolveIdFiles)).toMatchSnapshot('subResolveIdFiles')
    expect(resolveAbsPath(subLoadFiles, false)).toMatchSnapshot('subLoadFiles')
    loadFiles.length = 0
    subLoadFiles.length = 0
    const appJson = 'app.json.ts'// path.resolve(cwd, 'src/app.json.ts')
    expect(rootWatchFiles.includes(appJson)).toBe(true)
    // await touch(path.resolve(cwd, 'src/app.json.ts'))
    expect(resolveAbsPath(loadFiles, false)).toMatchSnapshot('loadFiles hmr')

    // expect(resolveAbsPath(subResolveIdFiles)).toMatchSnapshot('subResolveIdFiles')
    expect(resolveAbsPath(subLoadFiles, false)).toMatchSnapshot('subLoadFiles hmr')
    // expect(logger.success).toHaveBeenNthCalledWith(
    //   6,
    //   '已清空 dist/ 目录',
    // )
    // setImmediate(() => whyIsNodeRunning())
    ctx.watcherService.closeAll()
    // setImmediate(() => whyIsNodeRunning())
  })
})
