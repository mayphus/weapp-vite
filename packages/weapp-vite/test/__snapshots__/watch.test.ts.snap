// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`watch > watch > loadFiles hmr 1`] = `[]`;

exports[`watch > watch > subLoadFiles 1`] = `
[
  "packageB/pages/apple.js",
  "packageB/pages/banana.ts",
  "packageB/index.ts",
  "packageB/pages/banana.css?wxss",
  "utils/util.ts",
  "packageB/pages/apple.scss",
]
`;

exports[`watch > watch > subLoadFiles hmr 1`] = `[]`;

exports[`watch > watch > watchFiles 1`] = `
[
  "../package.json",
  "app.css",
  "app.json",
  "app.json.js",
  "app.json.ts",
  "app.less",
  "app.sass",
  "app.scss",
  "app.styl",
  "app.ts",
  "app.wxss",
  "components/Navbar/Navbar.css",
  "components/Navbar/Navbar.html",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.json.js",
  "components/Navbar/Navbar.json.ts",
  "components/Navbar/Navbar.less",
  "components/Navbar/Navbar.sass",
  "components/Navbar/Navbar.scss",
  "components/Navbar/Navbar.styl",
  "components/Navbar/Navbar.ts",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "packageA/index.css",
  "packageA/index.html",
  "packageA/index.js",
  "packageA/index.json",
  "packageA/index.json.js",
  "packageA/index.json.ts",
  "packageA/index.less",
  "packageA/index.sass",
  "packageA/index.scss",
  "packageA/index.styl",
  "packageA/index.wxml",
  "packageA/index.wxss",
  "packageA/pages/cat.css",
  "packageA/pages/cat.html",
  "packageA/pages/cat.json",
  "packageA/pages/cat.json.js",
  "packageA/pages/cat.json.ts",
  "packageA/pages/cat.less",
  "packageA/pages/cat.sass",
  "packageA/pages/cat.scss",
  "packageA/pages/cat.styl",
  "packageA/pages/cat.ts",
  "packageA/pages/cat.wxml",
  "packageA/pages/cat.wxss",
  "packageA/pages/dog.css",
  "packageA/pages/dog.html",
  "packageA/pages/dog.json",
  "packageA/pages/dog.json.js",
  "packageA/pages/dog.json.ts",
  "packageA/pages/dog.less",
  "packageA/pages/dog.sass",
  "packageA/pages/dog.scss",
  "packageA/pages/dog.styl",
  "packageA/pages/dog.ts",
  "packageA/pages/dog.wxml",
  "packageA/pages/dog.wxss",
  "pages/index/bbc.wxs",
  "pages/index/esm.wxs",
  "pages/index/index.css",
  "pages/index/index.html",
  "pages/index/index.json",
  "pages/index/index.json.js",
  "pages/index/index.json.ts",
  "pages/index/index.less",
  "pages/index/index.sass",
  "pages/index/index.scss",
  "pages/index/index.styl",
  "pages/index/index.ts",
  "pages/index/index.wxml",
  "pages/index/index.wxs.ts",
  "pages/index/index.wxss",
  "sitemap.json",
  "sitemap.json.js",
  "sitemap.json.ts",
  "stores/index.js",
  "theme.json",
  "theme.json.js",
  "theme.json.ts",
  "utils/util.ts",
]
`;

exports[`watch > watch > watchFiles-packageB 1`] = `
[
  "../package.json",
  "packageB/index.css",
  "packageB/index.html",
  "packageB/index.json",
  "packageB/index.json.js",
  "packageB/index.json.ts",
  "packageB/index.less",
  "packageB/index.sass",
  "packageB/index.scss",
  "packageB/index.styl",
  "packageB/index.ts",
  "packageB/index.wxml",
  "packageB/index.wxss",
  "packageB/pages/apple.css",
  "packageB/pages/apple.html",
  "packageB/pages/apple.js",
  "packageB/pages/apple.json",
  "packageB/pages/apple.json.js",
  "packageB/pages/apple.json.ts",
  "packageB/pages/apple.less",
  "packageB/pages/apple.sass",
  "packageB/pages/apple.scss",
  "packageB/pages/apple.styl",
  "packageB/pages/apple.wxml",
  "packageB/pages/apple.wxss",
  "packageB/pages/banana.css",
  "packageB/pages/banana.html",
  "packageB/pages/banana.json",
  "packageB/pages/banana.json.js",
  "packageB/pages/banana.json.ts",
  "packageB/pages/banana.less",
  "packageB/pages/banana.sass",
  "packageB/pages/banana.scss",
  "packageB/pages/banana.styl",
  "packageB/pages/banana.ts",
  "packageB/pages/banana.wxml",
  "packageB/pages/banana.wxss",
  "pages/index/bbc.wxs",
  "pages/index/cjs.wxs",
  "pages/index/dep.wxs",
  "pages/index/dep1.wxs",
  "pages/index/dep2.wxs",
  "pages/index/esm.wxs",
  "pages/index/index.wxs.ts",
  "pages/index/xx.wxs",
  "utils/util.ts",
]
`;

exports[`watch > watch > wxs 1`] = `
[
  {
    "code": "module.exports = {
  a: '1',
  v: '2'
};",
    "file": "pages/index/bbc.wxs",
  },
  {
    "code": "function tapName(event, ownerInstance) {
  console.log('tap 撒的撒打算', JSON.stringify(event));
}
module.exports = {
  tapName: tapName
};",
    "file": "pages/index/cjs.wxs",
  },
  {
    "code": "",
    "file": "pages/index/dep.wxs",
  },
  {
    "code": "",
    "file": "pages/index/dep1.wxs",
  },
  {
    "code": "",
    "file": "pages/index/dep2.wxs",
  },
  {
    "code": "var x = require('./cjs.wxs');
var xx = './dep.wxs';
var y = "./dep1.wxs";
var z = "./dep2" + '.wxs';
require(xx);
require(y);
require(z);
function tapName(event, ownerInstance) {
  console.log('tap 撒的撒打算', JSON.stringify(event));
}",
    "file": "pages/index/esm.wxs",
  },
  {
    "code": "module.exports.bar = bar;
module.exports.foo = void 0;
require("./xx.wxs");
var foo = module.exports.foo = '\\'hello world\\' from 我爱着';
function bar(d) {
  return d;
}",
    "file": "pages/index/index.wxs",
  },
  {
    "code": "console.log('xx');",
    "file": "pages/index/xx.wxs",
  },
]
`;

exports[`watch > watch 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "lodash-j7psf0cq.js",
  "miniprogram_npm/buffer/index.js",
  "miniprogram_npm/gm-crypto/index.js",
  "packageA/index.js",
  "packageA/pages/cat.js",
  "packageA/pages/cat.json",
  "packageA/pages/cat.wxml",
  "packageA/pages/cat.wxss",
  "packageA/pages/dog.js",
  "packageA/pages/dog.wxml",
  "packageA/pages/dog.wxss",
  "packageB/index.js",
  "packageB/miniprogram_npm/buffer/index.js",
  "packageB/miniprogram_npm/gm-crypto/index.js",
  "packageB/pages/apple.js",
  "packageB/pages/apple.json",
  "packageB/pages/apple.wxml",
  "packageB/pages/banana.js",
  "packageB/pages/banana.wxml",
  "packageB/util-b29wsyz4.js",
  "pages/index/bbc.wxs",
  "pages/index/cjs.wxs",
  "pages/index/dep.wxs",
  "pages/index/dep1.wxs",
  "pages/index/dep2.wxs",
  "pages/index/esm.wxs",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "pages/index/index.wxs",
  "pages/index/index.wxss",
  "pages/index/xx.wxs",
  "sitemap.json",
  "theme.json",
  "workers/hello.js",
  "workers/index.js",
  "workers/utils.js",
]
`;

exports[`watch > watch 2`] = `
Map {
  "packageB" => {
    "entries": [
      "packageB/pages/apple",
      "packageB/pages/banana",
      "packageB/index",
    ],
    "subPackage": {
      "entry": "index.js",
      "independent": true,
      "name": "pack2",
      "pages": [
        "pages/apple",
        "pages/banana",
      ],
      "root": "packageB",
    },
  },
}
`;
